polkadot/node/network/collator-protocol/src/collator_side/metrics.rs

```diff
@@ -116,8 +165,323 @@ impl metrics::Metrics for Metrics {
 				)?,
 				registry,
 			)?,
+			collation_fetch_latency: prometheus::register(
+				prometheus::Histogram::with_opts(
+					prometheus::HistogramOpts::new(
+						"polkadot_parachain_collation_fetch_latency",
+						"How much time collations spend waiting to be fetched",
+					)
+					.buckets(vec![
+						0.001, 0.01, 0.025, 0.05, 0.1, 0.15, 0.25, 0.35, 0.5, 0.75, 1.0, 2.0, 5.0,
+					]),
+				)?,
+				registry,
+			)?,
+			collation_backing_latency_time: prometheus::register(
+				prometheus::Histogram::with_opts(
+					prometheus::HistogramOpts::new(
+						"polkadot_parachain_collation_backing_latency_time",
+						"How much time it takes for a fetched collation to be backed",
+					)
+					.buckets(vec![
+						1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 8.0, 10.0, 12.0, 15.0, 18.0, 24.0, 30.0,
+					]),
+				)?,
+				registry,
+			)?,
+			collation_backing_latency: prometheus::register(
+				prometheus::Histogram::with_opts(
+					prometheus::HistogramOpts::new(
+						"polkadot_parachain_collation_backing_latency",
+						"How many blocks away from the relay parent are collations backed",
+					)
+					.buckets(vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]),
+				)?,
+				registry,
+			)?,
+			collation_inclusion_latency: prometheus::register(
+				prometheus::Histogram::with_opts(
+					prometheus::HistogramOpts::new(
+						"polkadot_parachain_collation_inclusion_latency",
+						"How many blocks it takes for a backed collation to be included",
+					)
+					.buckets(vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]),
+				)?,
+				registry,
+			)?,
+			collation_expired_total: prometheus::register(
+				prometheus::HistogramVec::new(
+					prometheus::HistogramOpts::new(
+						"polkadot_parachain_collation_expired",
+						"How many collations expired (not backed or not included)",
+					)
+					.buckets(vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]),
+					&["state"],
+				)?,
+				registry,
+			)?,
 		};
 
 		Ok(Metrics(Some(metrics)))
 	}
 }
+
+// Equal to allowed ancestry len + 1
+pub(crate) const MAX_BACKING_DELAY: BlockNumber = 3;
+// Paras availability period. In practice, candidates time out in exceptional situations.
+pub(crate) const MAX_AVAILABILITY_DELAY: BlockNumber = 10;
+
+// Collations are kept in the tracker, until they are included or expired
+#[derive(Default)]
+pub(crate) struct CollationTracker {
+	// Keep track of collation expiration block number.
+	expire: HashMap<BlockNumber, HashSet<Hash>>,
+	// All un-expired collation entries
+	entries: HashMap<Hash, CollationStats>,
```

Comment:
IMO candidate hash would be a better identifier here, because in theory candidate headers are not unique.
We can ignore this because the collators we're interested in observing are blockchains and don't have cycles

Reply:
Yeah, if same block gets rebuilt this for sure is gonna be a problem. It would work with candidate hash as well, we have the receipt on backed and inclusion and we can still match them.